_posts/2020_04_29_encrypted_keras_with_syft/encrypted_keras_with_syft.Rmd

```diff
@@ -0,0 +1,413 @@
+R process running in a console tab.
+
+To begin, we again define the model, two things being different now. First, for technical reasons, we need to pass in
+`batch_input_shape` instead of `input_shape`. Second, the final layer is "missing" the softmax activation. This is not an
+oversight -- SMPC `softmax` has not been implemented yet. (Depending on when you read this, that statement may no longer be
+true.) Were we training this model in *secret sharing* mode, this would of course be a problem; for classification though, all
+we care about is the maximum score.
+
+After model definition, we load the actual weights from the model we trained in the previous step. Then, the action begins. We
+create an ensemble of TFE workers that together run a distributed TensorFlow cluster. The model is *secret* *shared* with the
+workers, that is, model weights are split up into shares that, each inspected alone, are unusable. Finally, the model is
+*served*, i.e., made available to clients requesting predictions.
+
+How can a Keras model be *shared* and *served*? These are not methods provided by Keras itself. The magic is due to Syft
+*hooking* into Keras, extending the `model` object: cf. `hook <- sy$KerasHook(tf$keras)` in line 10.
+
+```{r}
+# serve.R
+# you could start R on the console and "source" this file
+
+# do this just once
+reticulate::py_install("syft[udacity]")
+
+library(tensorflow)
+library(keras)
+
+sy <- reticulate::import(("syft"))
+hook <- sy$KerasHook(tf$keras)
+
+batch_input_shape <- c(1, 28, 28, 1)
+
+model <- keras_model_sequential() %>%
+ layer_conv_2d(filters = 16, kernel_size = c(3, 3), batch_input_shape = batch_input_shape) %>%
+ layer_average_pooling_2d(pool_size = c(2, 2)) %>%
+ layer_activation("relu") %>%
+ layer_conv_2d(filters = 32, kernel_size = c(3, 3)) %>%
+ layer_average_pooling_2d(pool_size = c(2, 2)) %>%
+ layer_activation("relu") %>%
+ layer_conv_2d(filters = 64, kernel_size = c(3, 3)) %>%
+ layer_average_pooling_2d(pool_size = c(2, 2)) %>%
+ layer_activation("relu") %>%
+ layer_flatten() %>%
+ layer_dense(units = 10) 
+ 
+pre_trained_weights <- "model.hdf5"
+model$load_weights(pre_trained_weights)
+
+# create and start TFE cluster
+AUTO <- TRUE
+julia <- sy$TFEWorker(host = 'localhost:4000', auto_managed = AUTO)
+greg <- sy$TFEWorker(host = 'localhost:4001', auto_managed = AUTO)
+me <- sy$TFEWorker(host = 'localhost:4002', auto_managed = AUTO)
+cluster <- sy$TFECluster(julia, greg, me)
+cluster$start()
+
+# split up model weights into shares 
+model$share(cluster)
+
+# serve model (limiting number of requests)
+model$serve(num_requests = 3L)
+```
+
+Once the desired number of requests have been served, we can go to this R process, stop model sharing and shut down the
+cluster:
+
+```{r}
+# stop model sharing
+model$stop()
+
+# stop cluster
+cluster$stop()
+```
+
+Before that though, on to the client(s).
```

Comment:
Before what? Stopping the cluster?

Reply:
yeah... seems like i should rephrase that :-)